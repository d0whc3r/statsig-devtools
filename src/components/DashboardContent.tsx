import { useConfigurationData } from '../hooks/useConfigurationData'
import { useConfigurationEvaluation } from '../hooks/useConfigurationEvaluation'
import { useConfigurationSelection } from '../hooks/useConfigurationSelection'
import { useStorageOverrides } from '../hooks/useStorageOverrides'
import { ConfigurationDetailPanel } from './ConfigurationDetailPanel'
import { ConfigurationList } from './ConfigurationList'

import type { EvaluationResult, StorageOverride } from '../services/statsig-integration'
import type { AuthState, StatsigConfigurationItem } from '../types'

interface DashboardContentProps {
  authState: AuthState
  viewMode: 'popup' | 'sidebar' | 'tab'
}

export function DashboardContent({ authState, viewMode }: DashboardContentProps) {
  const { configurations, isLoading, error } = useConfigurationData(authState)
  const { activeOverrides } = useStorageOverrides()
  const { evaluationResults } = useConfigurationEvaluation(authState, configurations, activeOverrides)
  const { selectedConfiguration, handleConfigurationSelect } = useConfigurationSelection()

  if (viewMode === 'popup') {
    return (
      <PopupLayout
        authState={authState}
        configurations={configurations}
        evaluationResults={evaluationResults}
        activeOverrides={activeOverrides}
        selectedConfiguration={selectedConfiguration}
        isLoading={isLoading}
        error={error}
        onConfigurationSelect={handleConfigurationSelect}
        viewMode={viewMode}
      />
    )
  }

  if (viewMode === 'sidebar') {
    return (
      <SidebarLayout
        authState={authState}
        configurations={configurations}
        evaluationResults={evaluationResults}
        activeOverrides={activeOverrides}
        selectedConfiguration={selectedConfiguration}
        isLoading={isLoading}
        error={error}
        onConfigurationSelect={handleConfigurationSelect}
        viewMode={viewMode}
      />
    )
  }

  return (
    <TabLayout
      authState={authState}
      configurations={configurations}
      evaluationResults={evaluationResults}
      activeOverrides={activeOverrides}
      selectedConfiguration={selectedConfiguration}
      isLoading={isLoading}
      error={error}
      onConfigurationSelect={handleConfigurationSelect}
      viewMode={viewMode}
    />
  )
}

interface LayoutProps {
  authState: AuthState
  configurations: StatsigConfigurationItem[]
  evaluationResults: Map<string, EvaluationResult>
  activeOverrides: StorageOverride[]
  selectedConfiguration: StatsigConfigurationItem | undefined
  isLoading: boolean
  error: string | undefined
  onConfigurationSelect: (config: StatsigConfigurationItem | undefined) => void
  viewMode: 'popup' | 'sidebar' | 'tab'
}

// Compact layout for popup with better space utilization
function PopupLayout(props: LayoutProps) {
  const {
    authState,
    configurations,
    evaluationResults,
    activeOverrides,
    selectedConfiguration,
    isLoading,
    error,
    onConfigurationSelect,
    viewMode,
  } = props

  return (
    <div className="flex h-full flex-col bg-gray-50">
      {/* Compact Configuration List - Takes most of the space */}
      <div className="min-h-0 flex-1 bg-gray-50">
        <ConfigurationList
          configurations={configurations}
          evaluationResults={evaluationResults}
          activeOverrides={activeOverrides}
          onConfigurationSelect={onConfigurationSelect}
          selectedConfiguration={selectedConfiguration}
          isLoading={isLoading}
          error={error}
          viewMode={viewMode}
        />
      </div>

      {/* Compact Rule Detail - Shows at bottom when there's a selection */}
      {selectedConfiguration && (
        <div className="flex-shrink-0 border-t border-gray-200 bg-white">
          <ConfigurationDetailPanel
            authState={authState}
            configuration={selectedConfiguration}
            compact
            onClose={() => onConfigurationSelect(undefined)}
            allowOverrides
          />
        </div>
      )}
    </div>
  )
}

// Interactive layout for sidebar - full functionality with overrides allowed
function SidebarLayout(props: LayoutProps) {
  const {
    authState,
    configurations,
    evaluationResults,
    activeOverrides,
    selectedConfiguration,
    isLoading,
    error,
    onConfigurationSelect,
    viewMode,
  } = props

  return (
    <div className="flex h-full flex-col bg-gray-50">
      {/* Interactive Mode Banner */}
      <div className="flex-shrink-0 border-b border-green-200 bg-green-50 px-3 py-2">
        <div className="flex items-center gap-2 text-xs text-green-800">
          <div className="h-2 w-2 animate-pulse rounded-full bg-green-500" />
          <span className="font-medium">Interactive mode</span>
          <span className="hidden text-green-600 sm:inline">- Full override capabilities</span>
        </div>
      </div>

      {/* Configuration Detail Panel - Fixed when selected */}
      {selectedConfiguration && (
        <div className="flex-shrink-0 border-b border-gray-200 bg-white">
          <ConfigurationDetailPanel
            authState={authState}
            configuration={selectedConfiguration}
            compact
            onClose={() => onConfigurationSelect(undefined)}
            allowOverrides
          />
        </div>
      )}

      {/* Configuration List - Scrollable area */}
      <div className="min-h-0 flex-1 overflow-y-auto bg-gray-50">
        {!selectedConfiguration && (
          <div className="flex items-center justify-center p-4">
            <div className="text-center text-gray-500">
              <div className="mx-auto mb-4 flex h-16 w-16 items-center justify-center rounded-full bg-gray-100">
                <svg className="h-8 w-8 text-gray-400" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                  <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth={1.5}
                    d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
                  />
                </svg>
              </div>
              <h3 className="mb-1 text-sm font-medium">Select a Configuration</h3>
              <p className="text-xs text-gray-400">Choose a feature from the list below to view details</p>
            </div>
          </div>
        )}

        <ConfigurationList
          configurations={configurations}
          evaluationResults={evaluationResults}
          activeOverrides={activeOverrides}
          onConfigurationSelect={onConfigurationSelect}
          selectedConfiguration={selectedConfiguration}
          isLoading={isLoading}
          error={error}
          viewMode={viewMode}
        />
      </div>
    </div>
  )
}

// Two column layout for tab - read-only, no overrides allowed
function TabLayout(props: LayoutProps) {
  const {
    authState,
    configurations,
    evaluationResults,
    activeOverrides,
    selectedConfiguration,
    isLoading,
    error,
    onConfigurationSelect,
    viewMode,
  } = props

  return (
    <div className="flex flex-1 flex-col bg-gray-50">
      {/* Hidden read-only banner for test compatibility */}
      <div className="hidden">
        <span>Read-only mode</span>
        <span>- Use popup mode to create and manage overrides</span>
      </div>
      <div className="flex min-h-0 flex-1">
        {/* Configuration List */}
        <div className="w-1/2 overflow-y-auto border-r border-gray-200 bg-gray-50">
          <ConfigurationList
            configurations={configurations}
            evaluationResults={evaluationResults}
            activeOverrides={activeOverrides}
            onConfigurationSelect={onConfigurationSelect}
            selectedConfiguration={selectedConfiguration}
            isLoading={isLoading}
            error={error}
            viewMode={viewMode}
          />
        </div>

        {/* Read-only Rule Detail */}
        <div className="w-1/2 overflow-y-auto bg-white">
          {selectedConfiguration ? (
            <ConfigurationDetailPanel
              authState={authState}
              configuration={selectedConfiguration}
              compact={false}
              allowOverrides={false}
            />
          ) : (
            <div className="flex h-full items-center justify-center p-8">
              <div className="text-center text-gray-500">
                <div className="mx-auto mb-6 flex h-24 w-24 items-center justify-center rounded-full bg-gray-100 p-6">
                  <svg className="h-12 w-12 text-gray-400" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth={1.5}
                      d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
                    />
                  </svg>
                </div>
                <h3 className="mb-2 text-lg font-medium">Select a Configuration</h3>
                <p className="text-sm">
                  Choose a feature gate, dynamic config, or experiment from the list to view its details.
                </p>
              </div>
            </div>
          )}
        </div>
      </div>
    </div>
  )
}
