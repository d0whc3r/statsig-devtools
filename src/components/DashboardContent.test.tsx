import { describe, expect, it, vi } from 'vitest'

import { DashboardContent } from './DashboardContent'

import type { AuthState } from '../types'

import { render, screen } from '@testing-library/react'

// Mock hooks
vi.mock('../hooks/useConfigurationData', () => ({
  useConfigurationData: vi.fn(() => ({
    configurations: [],
    isLoading: false,
    error: undefined,
  })),
}))

vi.mock('../hooks/useConfigurationEvaluation', () => ({
  useConfigurationEvaluation: vi.fn(() => ({
    evaluationResults: new Map(),
  })),
}))

vi.mock('../hooks/useConfigurationSelection', () => ({
  useConfigurationSelection: vi.fn(() => ({
    selectedConfiguration: undefined,
    handleConfigurationSelect: vi.fn(),
  })),
}))

vi.mock('../hooks/useStorageOverrides', () => ({
  useStorageOverrides: vi.fn(() => ({
    activeOverrides: [],
  })),
}))

// Mock child components
vi.mock('./ConfigurationList', () => ({
  ConfigurationList: ({ viewMode }: { viewMode: string }) => (
    <div data-testid="configuration-list" data-view-mode={viewMode}>
      Configuration List
    </div>
  ),
}))

vi.mock('./ConfigurationDetailPanel', () => ({
  ConfigurationDetailPanel: ({ allowOverrides }: { allowOverrides: boolean }) => (
    <div data-testid="configuration-detail-panel" data-allow-overrides={allowOverrides}>
      Configuration Detail Panel
    </div>
  ),
}))

describe('DashboardContent', () => {
  const mockAuthState: AuthState = {
    isAuthenticated: true,
    isLoading: false,
    consoleApiKey: 'test-client-key',
    projectName: 'test-project',
    error: undefined,
  }

  it('should match snapshot with popup view mode', () => {
    const { container } = render(<DashboardContent authState={mockAuthState} viewMode="popup" />)
    expect(container).toMatchSnapshot()
  })

  it('should match snapshot with tab view mode', () => {
    const { container } = render(<DashboardContent authState={mockAuthState} viewMode="tab" />)
    expect(container).toMatchSnapshot()
  })

  it('should match snapshot with sidebar view mode', () => {
    const { container } = render(<DashboardContent authState={mockAuthState} viewMode="sidebar" />)
    expect(container).toMatchSnapshot()
  })

  describe('View Mode Restrictions', () => {
    it('should allow overrides in popup mode', () => {
      render(<DashboardContent authState={mockAuthState} viewMode="popup" />)

      // In popup mode, overrides should be allowed
      const configList = screen.getByTestId('configuration-list')
      expect(configList).toHaveAttribute('data-view-mode', 'popup')
    })

    it('should show read-only banner and disable overrides in tab mode', () => {
      render(<DashboardContent authState={mockAuthState} viewMode="tab" />)

      // Should show read-only banner
      expect(screen.getByText('Read-only mode')).toBeInTheDocument()
      expect(screen.getByText('- Use popup mode to create and manage overrides')).toBeInTheDocument()

      // Configuration list should be in tab mode
      const configList = screen.getByTestId('configuration-list')
      expect(configList).toHaveAttribute('data-view-mode', 'tab')
    })

    // Note: Testing allowOverrides prop passing would require more complex mocking
    // The core functionality is tested through the read-only banners and layout tests
  })

  describe('Layout Differences', () => {
    it('should use single column layout in popup mode', () => {
      render(<DashboardContent authState={mockAuthState} viewMode="popup" />)

      // Should not have the two-column layout class
      expect(screen.queryByText('Select a Configuration')).not.toBeInTheDocument()
    })

    it('should use single column layout in sidebar mode', () => {
      render(<DashboardContent authState={mockAuthState} viewMode="sidebar" />)

      // Should not have the two-column layout class
      expect(screen.getByText('Select a Configuration')).toBeInTheDocument()
    })

    it('should use two column layout in tab mode', () => {
      render(<DashboardContent authState={mockAuthState} viewMode="tab" />)

      // Should show the empty state message for the right column
      expect(screen.getByText('Select a Configuration')).toBeInTheDocument()
      expect(
        screen.getByText('Choose a feature gate, dynamic config, or experiment from the list to view its details.'),
      ).toBeInTheDocument()
    })
  })
})
