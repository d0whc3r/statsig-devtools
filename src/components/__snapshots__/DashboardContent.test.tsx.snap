// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DashboardContent > should match snapshot with popup view mode 1`] = `
<div>
  <div
    class="flex h-full flex-col bg-gray-50"
  >
    <div
      class="min-h-0 flex-1 bg-gray-50"
    >
      <div
        data-testid="configuration-list"
        data-view-mode="popup"
      >
        Configuration List
      </div>
    </div>
  </div>
</div>
`;

exports[`DashboardContent > should match snapshot with sidebar view mode 1`] = `
<div>
  <div
    class="flex h-full flex-col bg-gray-50"
  >
    <div
      class="flex-shrink-0 border-b border-green-200 bg-green-50 px-3 py-2"
    >
      <div
        class="flex items-center gap-2 text-xs text-green-800"
      >
        <div
          class="h-2 w-2 animate-pulse rounded-full bg-green-500"
        />
        <span
          class="font-medium"
        >
          Interactive mode
        </span>
        <span
          class="hidden text-green-600 sm:inline"
        >
          - Full override capabilities
        </span>
      </div>
    </div>
    <div
      class="min-h-0 flex-1 overflow-y-auto bg-gray-50"
    >
      <div
        class="flex items-center justify-center p-4"
      >
        <div
          class="text-center text-gray-500"
        >
          <div
            class="mx-auto mb-4 flex h-16 w-16 items-center justify-center rounded-full bg-gray-100"
          >
            <svg
              class="h-8 w-8 text-gray-400"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
            >
              <path
                d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              />
            </svg>
          </div>
          <h3
            class="mb-1 text-sm font-medium"
          >
            Select a Configuration
          </h3>
          <p
            class="text-xs text-gray-400"
          >
            Choose a feature from the list below to view details
          </p>
        </div>
      </div>
      <div
        data-testid="configuration-list"
        data-view-mode="sidebar"
      >
        Configuration List
      </div>
    </div>
  </div>
</div>
`;

exports[`DashboardContent > should match snapshot with tab view mode 1`] = `
<div>
  <div
    class="flex flex-1 flex-col bg-gray-50"
  >
    <div
      class="hidden"
    >
      <span>
        Read-only mode
      </span>
      <span>
        - Use popup mode to create and manage overrides
      </span>
    </div>
    <div
      class="flex min-h-0 flex-1"
    >
      <div
        class="w-1/2 overflow-y-auto border-r border-gray-200 bg-gray-50"
      >
        <div
          data-testid="configuration-list"
          data-view-mode="tab"
        >
          Configuration List
        </div>
      </div>
      <div
        class="w-1/2 overflow-y-auto bg-white"
      >
        <div
          class="flex h-full items-center justify-center p-8"
        >
          <div
            class="text-center text-gray-500"
          >
            <div
              class="mx-auto mb-6 flex h-24 w-24 items-center justify-center rounded-full bg-gray-100 p-6"
            >
              <svg
                class="h-12 w-12 text-gray-400"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="1.5"
                />
              </svg>
            </div>
            <h3
              class="mb-2 text-lg font-medium"
            >
              Select a Configuration
            </h3>
            <p
              class="text-sm"
            >
              Choose a feature gate, dynamic config, or experiment from the list to view its details.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
